pdns/pdns_recursor.cc

```diff
@@ -1720,6 +1712,14 @@ void indirectDoProcessUDPQuestion(void *p)
 
 string *doProcessUDPQuestion(const std::string& question, const ComboAddress& fromaddr, const ComboAddress& destaddr, struct timeval tv, int fd)
 {
+  if(MT->numProcesses() > g_maxMThreads) {
+    if(!g_quiet)
+      L<<Logger::Notice<<t_id<<" ["<<MT->getTid()<<"/"<<MT->numProcesses()<<"] DROPPED question from "<<fromaddr.toStringWithPort()<<", over capacity"<<endl;
+
+    g_stats.overCapacityDrops++;
+    return 0;
+  }
```

Comment:
This does however prevent serving a response from the packet cache even if we already have too many mthreads. Perhaps you can only check it here if `start-mthread-early` is set, and skip the check in `realDoProcessUDPQuestion()` if `inmthread` is true?